src/core/emitter.ts

```diff
@@ -0,0 +1,78 @@
+import { EventEmitter } from 'eventemitter3';
+
+class Emitter extends EventEmitter {
+  _barriers: Barriers;
+  _lookups: Lookups;
+
+  constructor() {
+    super();
+    this._barriers = {};
+    this._lookups = {};
+  }
+
+  emit(event: string, ...args: any[]): boolean {
+    const result = super.emit(event, ...args);
+    const keys = this._lookups[event] || [];
+
+    keys.forEach((key) => {
+      const inc = this._barriers[key].events[event] + 1;
+      this._barriers[key].events = { ...this._barriers[key].events, [event]: inc };
+
+      const shouldInvoke = Object.keys(this._barriers[key].events).every(ev => this._barriers[key].events[ev] >= 1);
+
+      if (shouldInvoke) {
+        this._barriers[key].cb.forEach(cb => cb());
```

Comment:
Ensure that callbacks are invoked in the correct context (see: `eventemitter3` for reference).